### ISSUE TITLE
"multi-character character constant" warning is not displayed for 4-character constants

### ISSUE THREAD

**Issue by strimo378** [llvm/llvm-project#137471]
Hi all,

I found an inconsistency in the warning of multi-character constants. In the following code, the first 2 cases produce a warning while the third one does not:

```
int x2 = '\x12\x34';
int x3 = '\x12\x34\x56';
int x4 = '\x12\x34\x56\x78';
```

This code outputs

```
<source>:3:10: warning: multi-character character constant [-Wmultichar]
    3 | int x2 = '\x12\x34';
      |          ^
<source>:4:10: warning: multi-character character constant [-Wmultichar]
    4 | int x3 = '\x12\x34\x56';
```

See compiler explorer: https://godbolt.org/z/Ehrzjqjxh

gcc is showing 3 warnings.

**Comment by svs-quic** [comment]
I found this patch https://reviews.llvm.org/D87962 from which it seems that the warning is different (-Wfour-char-constants) and ignored by default. If I add -Wfour-char-constants to the command line explicitly I see the warning.

https://godbolt.org/z/xfGGdYKW8

**Comment by strimo378** [comment]
Thank you for providing the link to the patch.

I see the point of having a different warning for 4-char constants. It is hard to understand that this warning is not active when the general multichar warning is active. And it is very hard to activate the warning since it is in no other group, i.e.,-Wall is not enough.

Maybe it would make sense to change the text of the general warning to something like "2- or 3-character constant", or give a hint to -Wfour-char-constants.

Or remove DefaultIgnore from warn_four_char_character_literal and make it a subgroup of MultiChar.

**Comment by AaronBallman** [comment]
> I see the point of having a different warning for 4-char constants. It is hard to understand that this warning is not active when the general multichar warning is active. And it is very hard to activate the warning since it is in no other group, i.e.,-Wall is not enough.

Four-char codes were a common coding pattern, particularly for folks coming from Apple Classic backgrounds, so it's a separate diagnostic group that is off by default because of how common the code is in the wild: https://en.wikipedia.org/wiki/FourCC

> Maybe it would make sense to change the text of the general warning to something like "2- or 3-character constant", or give a hint to -Wfour-char-constants.
> 
> Or remove DefaultIgnore from warn_four_char_character_literal and make it a subgroup of MultiChar.

It might be more plausible these days to group it under `MultiChar`; that one is under `-Wmost` which means the diagnostic will trigger more often, but the use of four char codes fell out of style 20 years ago.

However, the current behavior is by-design.

**Comment by cor3ntin** [comment]
@AaronBallman I think we should align on GCC here  - It makes sense to retain a separate flag, but to switch its default. While these were at some point used - and sometimes still are, they do cause confusion for a lot of folks (and were always just... _clever_)

**Comment by AaronBallman** [comment]
> While these were at some point used - and sometimes still are, they do cause confusion for a lot of folks (and were always just... _clever_)

The trouble is, there are major libraries which still use these, such as QuickTime, wxWidgets, various file formats (especially as a "magic code" to identify the file type), etc. So "sometimes" is load-bearing. I think moving it to `-Wmost` first (by putting it under `-Wmultichar`) would be a conservative step in the right direction. If that goes over well, then we can consider making it on by default as the next step. The alternative is a bit more aggressive (it may still be okay, but I generally prefer we move conservatively unless the changes improve security posture significantly or there's some other pressure driving the change).
